ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java

```diff
@@ -238,9 +243,29 @@
         routerOperation, successTarget, parallelism, originatingDcNotFoundFailureThreshold);
   }
 
+  /**
+   * The dynamic success target is introduced mainly for following use case:
+   * In the intermediate state of "move replica", when decommission of old replicas is initiated(but hasn't transited to
+   * INACTIVE yet), the PUT requests should be rejected on old replicas. For frontends, they are seeing both old and new
+   * replicas(lets say 3 old and 3 new) and the success target should be 6 - 1 = 5. In the aforementioned scenario, PUT
+   * request failed on 3 old replicas. It seems we should fail whole PUT operation because number of remaining requests
+   * is already less than success target.
+   * From another point of view, however, PUT request is highly likely to succeed on 3 new replicas and we actually
+   * could consider it success without generating "slip put" (which makes PUT latency worse). The reason is, if new PUTs
+   * already succeeded on at least 2 new replicas,  read-after-write should always succeed because frontends are always
+   * able to see new replicas and subsequent READ/DELETE/TtlUpdate request should succeed on at least 2 aforementioned
+   * new replicas.
+   */
   @Override
   public boolean hasSucceeded() {
-    return succeededCount >= successTarget;
+    boolean hasSucceeded;
+    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget) {
+      hasSucceeded =
+          succeededCount >= Math.max(totalReplicaCount - disabledCount - 1, routerConfig.routerPutSuccessTarget);
```

Comment:
i think we could use `successTarget` in these places

Reply:
Unfortunately, we can't. Here is a counter example:  
 We have 6 replicas(3 old, 3 new) and they are all eligible from frontend's view.  `successTarget` is now 5 (= 6 - 1). Meanwhile, decommission is initiated on 3 old replicas and PutOperation gets 3 success (from new replicas) and 3 disabled result. We should consider this a success. However, if we use `succeededCount >= Math.max(totalReplicaCount - disabledCount - 1, successTarget)`, then `succeededCount` is less than `successTarget` (3 < 5) in this case, which make the operation fail.